File: UtrechtJUG/KotlinMigration/src/main/java/nl/blue4it/basic/service/PaymentService.java
package nl.blue4it.basic.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.blue4it.basic.controller.dto.request.Payment;
import nl.blue4it.basic.service.async.AsyncSender;
import nl.blue4it.basic.service.exception.FutureException;
import nl.blue4it.basic.service.exception.PaymentRejectedException;
import nl.blue4it.basic.service.exception.SomeOtherClientException;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
@RequiredArgsConstructor
@Slf4j
public class PaymentService {
    private final AsyncSender client;

    public void processPayment(List<Payment> payments) throws PaymentRejectedException {
        // no stream needed
        payments.stream()
                .map(this::mapPayment)
                .forEach(payment -> {
                    try {
                        client.send(payment);
                    } catch (FutureException | SomeOtherClientException e) {
                        // catch problem
                        throw new PaymentRejectedException("");
                    }
                });
    }

    private example.avro.Payment mapPayment(Payment payment) {
        int durationNumber;

        log.info("Going to map with  {}", payment.getDuration());

        // can change to switch to when
        if (payment.getDuration().equals("weekly")) {
            durationNumber = 1;
        } else if (payment.getDuration().equals("monthly")) {
            durationNumber = 2;
        } else if (payment.getDuration().equals("yearly")) {
            durationNumber = 3;
        } else {
            durationNumber = 0;
        }

        // type alias
        return new example.avro.Payment(payment.getName(), payment.getFromIban(), payment.getToIban(),
                payment.getAmount(), payment.getBalance(), true, durationNumber);
    }
}

File: UtrechtJUG/KotlinMigration/src/main/kotlin/nl/blue4it/basic/service/PaymentService.kt
package nl.blue4it.basic.service

import nl.blue4it.basic.controller.dto.request.Payment
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

typealias RequestPayment = Payment
typealias AvroPayment = example.avro.Payment

@Component
class PaymentService constructor(private val client: AsyncSender) {

    companion object {
        val log: Logger = LoggerFactory.getLogger(PaymentService::class.java)
    }

    fun processPayment(payments: List<Payment>) {
        payments
            .map { payment -> payment.mapPayment() }
            .forEach { payment: AvroPayment ->
                kotlin.runCatching {
                    client.send(payment)
                }.onFailure {
                    log.error("something bad happened")
                }
            }
    }

    private fun RequestPayment.mapPayment(): AvroPayment {
        var durationNumber = 0

        this.duration?.let {
            log.info("Going to map with  {}", this.duration)

            durationNumber = when(this.duration) {
                "weekly" -> 1
                "monthly" -> 2
                "yearly" -> 1
                else -> 0
            }
        }.also {
            // create metric here
        } ?: run {
            log.error("something is going wrong")
        }

        return AvroPayment(
            this.name, this.fromIban, this.toIban,
            this.amount, this.balance, true, durationNumber
        )
    }
}